src/Razor/Microsoft.AspNetCore.Razor.Language/src/RazorSourceDocument.cs

```diff
@@ -193,6 +193,11 @@ public static RazorSourceDocument ReadFrom(RazorProjectItem projectItem)
                 filePath = projectItem.FilePath;
             }
 
+            if (projectItem.SourceDocument is not null)
```

Comment:
Why not just re-create a source document?

Reply:
This and the change above are designed to work in conjunction with some work we've done to support source generators.

We have a new `SourceGeneratorProjectItem` ([ref](https://github.com/dotnet/sdk/blob/safia/rsg/src/RazorSdk/SourceGenerators/SourceGeneratorProjectItem.cs)) type that gets constructed from the list of files that Roslyn provides in a generator's compilation context ([ref](https://github.com/dotnet/sdk/blob/safia/rsg/src/RazorSdk/SourceGenerators/RazorSourceGenerationContext.cs#L105)).

These additional files are provided as `SourceText`s. We have a `SourceTextRazorSourceDocument` type ([ref](https://github.com/dotnet/sdk/blob/safia/rsg/src/RazorSdk/SourceGenerators/SourceTextRazorSourceDocument.cs) that we create from the `SourceText`.

Here we are updating the `ReadFrom` method to directly use the `SourceDocument` that we compute from the `SourceText` provided by Roslyn instead of reading in the file contents from a stream.

You can view a hacky workaround I used to get around this [here](https://github.com/dotnet/sdk/blob/safia/rsg/src/RazorSdk/SourceGenerators/SourceGeneratorProjectItem.cs#L43) by reading the string of the `SourceText` into a stream but this isn't ideal.

Hope this explanation was clear. Thoughts?

Reply:
Thanks for the clarification, definitely helps piece it all together a bit more.  It's been a while since I've dug through the code but why can't we have the project item read directly from the Roslyn source text?

Reply:
We did create a ProjectItem implementation that reads directly from the Roslyn source text ([ref](https://github.com/dotnet/sdk/blob/safia/rsg/src/RazorSdk/SourceGenerators/SourceGeneratorProjectItem.cs#L10)). But currently, the project engine computes the source document ([ref](https://github.com/dotnet/aspnetcore/blob/70a880f30cd1537ff3777f51efe18469c8da68db/src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectEngine.cs#L78)) from a project item by reading from a stream ([ref](https://github.com/dotnet/aspnetcore/blob/70a880f30cd1537ff3777f51efe18469c8da68db/src/Razor/Microsoft.AspNetCore.Razor.Language/src/RazorSourceDocument.cs#L196-L201)).

We can try implementing an overload of `ReadFrom` that takes in a `SourceGeneratorProjectItem`

```csharp
public static RazorSourceDocument ReadFrom(SourceGeneratorProjectItem projectItem)
{
   return new SourceGeneratorSourceDocument(projectItem.AdditionalText); // or similiar
}
```

But that would add source generator-specific logic in the compiler.

An alternative is to create some sort of abstraction where in a custom `ProjectItem` can define how source documents should be read from it similar to @pranavkm's recommendation above.

Reply:
> An alternative is to create some sort of abstraction where in a custom `ProjectItem` can define how source documents should be read from it similar to @pranavkm's recommendation above.

Ya I think that'd be better. It'd be a more defined route of translating from one type to another.

Reply:
Yep! Will create an issue to track this. We want to get the initial E2E with source generators working then refine things.